storage/cleanup_prop_test.go

```diff
@@ -0,0 +1,283 @@
+			cm := propTestCleanupMgr(ctrl, cRopts, gNamespaces(nses).asDatabaseNamespace()...)
+			_, cleanupTimes := cm.commitLogTimes(t)
+			for _, ct := range cleanupTimes {
+				for _, ns := range nses {
+					s, e := commitLogNamespaceBlockTimes(ct, cRopts.BlockSize(), ns.Options().RetentionOptions())
+					if ns.NeedsFlush(s, e) {
+						return false, fmt.Errorf("trying to cleanup commit log at %v, but ns needsFlush; (range: %v, %v)",
+							ct.String(), s.String(), e.String())
+					}
+				}
+			}
+			return true, nil
+		},
+		gen.TimeRange(now.Add(-year), 2*year).WithLabel("cleanup time"),
+		genCommitLogRetention().WithLabel("commit log retention"),
+		gen.SliceOfN(3, genNamespace(now)).WithLabel("namespaces"),
+	))
+
+	properties.TestingRun(t)
+}
+
+type gNamespaces []*gNamespace
+
+func (n gNamespaces) asDatabaseNamespace() []databaseNamespace {
+	nses := make([]databaseNamespace, 0, len(n))
+	for _, ns := range n {
+		nses = append(nses, ns)
+	}
+	return nses
+}
+
+// generated namespace struct
+type gNamespace struct {
+	databaseNamespace
+
+	opts              namespace.Options
+	ropts             *gRetention
+	blockSize         time.Duration
+	oldestBlock       time.Time
+	newestBlock       time.Time
+	needsFlushMarkers []bool
+}
+
+func (ns *gNamespace) String() string {
+	var buf bytes.Buffer
+	buf.WriteString(fmt.Sprintf("\n\tretention: %v", ns.ropts.String()))
+
+	buf.WriteString(fmt.Sprintf("\n\tneedsFlush: \n"))
+	w := new(tabwriter.Writer)
+	w.Init(&buf, 5, 0, 1, ' ', 0)
+	fmt.Fprintln(w, "blockStart\tneedFlush\t")
+	for i := range ns.needsFlushMarkers {
+		t := ns.oldestBlock.Add(time.Duration(i) * ns.blockSize)
+		fmt.Fprintf(w, "%v\t%v\t\n", t, ns.needsFlushMarkers[i])
+	}
+	w.Flush()
+
+	return buf.String()
+}
+
+func (ns *gNamespace) blockIdx(t time.Time) int {
+	idx := int(t.Truncate(ns.blockSize).Sub(ns.oldestBlock) / ns.blockSize)
+	if idx < 0 {
+		return 0
+	}
+	if idx >= len(ns.needsFlushMarkers) {
+		return len(ns.needsFlushMarkers) - 1
+	}
+	return idx
+}
+
+func (ns *gNamespace) Options() namespace.Options {
+	return ns.opts
+}
+
+func (ns *gNamespace) NeedsFlush(start, end time.Time) bool {
+	if start.Before(ns.oldestBlock) && end.Before(ns.oldestBlock) {
+		return false
+	}
+	if start.After(ns.newestBlock) && end.After(ns.newestBlock) {
+		return false
+	}
+	sIdx, eIdx := ns.blockIdx(start), ns.blockIdx(end)
+	for i := sIdx; i <= eIdx; i++ {
+		if ns.needsFlushMarkers[i] {
+			return true
+		}
+	}
+	return false
+}
+
+// generator for gNamespace
+func genNamespace(t time.Time) gopter.Gen {
+	return func(genParams *gopter.GenParameters) *gopter.GenResult {
+		rng := genParams.Rng
+		ropts := randomRetention(rng)
+		oldest := retention.FlushTimeStart(ropts, t)
+		newest := retention.FlushTimeEnd(ropts, t)
+
+		n := numIntervals(oldest, newest, ropts.BlockSize())
+		flushStates := make([]bool, n)
+		for i := range flushStates {
+			flushStates[i] = rng.Float32() > 0.6 // flip a coin to get a bool
```

Comment:
Yep, I set this empirically after a few runs trying to maximise the number of distinct commit log times to be cleaned up.